src/api/typed.rs

```diff
@@ -41,6 +42,16 @@ pub struct Api<K> {
     pub(in crate::api) phantom: PhantomData<K>,
 }
 
+pub trait LogOperation {
+    fn log_operation(&self, name: &str, lp: &LogParams) -> Result<String>;
+}
+
+pub trait Log: LogOperation {
+    fn log(&self, name: &str, lp: &LogParams) -> Result<String> {
+        Ok(self.log_operation(name, lp)?)
+    }
+}
```

Comment:
This is a nice setup between private and public methods that looks like it allows you to very easily implement the Log trait. Cool pattern. Unfortunately, if you follow do my suggestion, this won't be needed.